modules/FvwmEvent/FvwmEvent.1
.\" t
.\" @(#)FvwmEvent.1	3/28/94
.TH FvwmEvent 1 "25 April 2002" FVWM "FVWM Modules"
.UC
.SH NAME
\fIFvwmEvent\fP \- the FVWM Event module
.SH SYNOPSIS
\fIFvwmEvent\fP is a more versatile replacement for \fIFvwmAudio\fP.
It can in general be used to hook any \fIfvwm\fP function or program to any
window manager event. E.g: Delete unwanted Netscape Pop ups or
application error pop ups as they appear, play sounds, log events to a
file and the like. Be creative, You'll find a use for it.

\fIFvwmEvent\fP is spawned by \fIfvwm\fP, so no command line invocation will
work.  From within the \fI.fvwm2rc\fP file, \fIFvwmEvent\fP is spawned as
follows:
.nf
.sp
Module FvwmEvent
.sp
.fi
or from within an \fIfvwm\fP pop-up menu:
.nf
.sp
DestroyMenu Module-Popup
AddToMenu Module-Popup "Modules" Title
+ "Event"        Module FvwmEvent
+ "Auto"         Module FvwmAuto 200
+ "Buttons"      Module FvwmButtons
+ "Console"      Module FvwmConsole
+ "Ident"        Module FvwmIdent
+ "Banner"       Module FvwmBanner
+ "Pager"        Module FvwmPager 0 3
.sp
.fi
.SH DESCRIPTION
The \fIFvwmEvent\fP module communicates with the \fIfvwm\fP window manager
to bind \fIactions\fP to window manager \fIevents\fP.  Different actions
may be assigned to distinct window manager events.

\fIFvwmEvent\fP can be used to bind sound files to events like
\fIFvwmAudio\fP (RiP) did. It can be used for logging event traces to
a log file, while debugging \fIfvwm\fP.

\fIFvwmEvent\fP can also have builtin support for the rplay library.
(heritage of FvwmAudio)

.SH INVOCATION
The invocation method was shown in the synopsis section. No command
line invocation is possible. \fIFvwmEvent\fP must be invoked by the
\fIfvwm\fP window manager. \fIFvwmEvent\fP accepts a single
argument:

.IP \-audio
Enables FvwmAudio compatibility mode.

.IP alias
Makes FvwmEvent use \fIalias\fP as its name. This affects which lines
from the user's configuration file are used.

Invoking FvwmEvent as \fIFvwmAudio\fP (either by using an alias or
creating a symlink) enables FvwmAudio compatibility mode.

.sp

.SH CONFIGURATION OPTIONS
\fIFvwmEvent\fP gets config info from \fBfvwm\fP's module configuration
database (see
.IR fvwm (1),
section
.BR "MODULE COMMANDS" ),
and looks for certain configuration options:

.IP "*FvwmEvent: Cmd \fIcommand\fP"
This determines the \fIfvwm\fP function that is to be called with the
event parameters. You might want to do one of the following (details below):
.nf
.sp
	# play sounds
	*FvwmEvent: Cmd \fIbuiltin-rplay\fP

	# execute distinct fvwm functions
	*FvwmEvent: Cmd

	# execute distinct external programs
	*FvwmEvent: Cmd exec
.sp
.fi
This version of \fIFvwmEvent\fP has builtin \fIrplay\fP support which does not
need to invoke an external audio player to play sounds.  The rplay
support is enabled when \fIFvwmEvent\fP is compiled with \fIHAVE_RPLAY\fP
defined and when \fIFvwmEvent: Cmd\fP is set to \fIbuiltin-rplay\fP. See
remarks below if \fIFvwmEvent\fP is invoked in FvwmAudio compatibility mode.

For example:
.nf
.sp
	*FvwmEvent: Cmd \fIbuiltin-rplay\fP
	*FvwmEvent: add_window drip.au
.sp
.fi
rplay can be obtained via anonymous ftp at
.nf
.sp
	<URL:ftp://ftp.sdsu.edu/pub/rplay> or
	<URL:ftp://ftp.x.org/contrib/Event/audio/rplay>
.sp
.fi
\fIFvwmEvent\fP also has support for any other external program.
e.g: the rsynth 'say' command:
.nf
.sp
	*FvwmEvent: Cmd "Exec /rsynth/say"
	*FvwmEvent: destroy_window "window closed"
.sp
.fi
You can also use \fIfvwm\fP's builtin \fIEcho\fP command as
\fIFvwmEvent: Cmd\fP to obtain debug output for \fIfvwm\fP events quietly.
I used this setup to debug FvwmAuto:
.nf
.sp
	*FvwmEvent: Cmd \fIEcho\fP
	*FvwmEvent: focus_change "focus change"
	*FvwmEvent: raise_window "raise window"
.sp
.fi
You can even call different shell commands for each event just by setting
.nf
.sp
	*FvwmEvent: Cmd exec
	*FvwmEvent: add_window killname "APPL ERROR"
.sp
.fi
.IP "*FvwmEvent: PassID"
Specifies that the event action will have an ID parameter added to the end
of the command line. Most events will have the windowID of the window that the
event refers to, new_desk will have the new desk number. The windowID is a
hexadecimal string preceded by 0x, desk numbers are decimal.

.IP "*FvwmEvent: \fIwindow-manager-event action-or-filename\fP"
Binds particular actions to window manager events.

e.g. for audio-events:
.nf
.sp
	*FvwmEvent: startup TaDa.au
	*FvwmEvent: shutdown Elvis_Left.au
	*FvwmEvent: unknown doh.au

	*FvwmEvent: new_page beam_trek.au
	*FvwmEvent: new_desk beam_trek.au
	*FvwmEvent: add_window drip.au
	*FvwmEvent: raise_window swoosh.au
	*FvwmEvent: lower_window swoosh.au
	*FvwmEvent: configure_window hammer.au
	*FvwmEvent: focus_change boing.au
	*FvwmEvent: enter_window boing.au
	*FvwmEvent: leave_window boing.au
	*FvwmEvent: destroy_window explosion.au
	*FvwmEvent: iconify ploop.au
	*FvwmEvent: deiconify ploop.au
	*FvwmEvent: window_name huh.au
	*FvwmEvent: icon_name beep.au
	*FvwmEvent: visible_icon_name beep.au
	*FvwmEvent: res_class beep.au
	*FvwmEvent: res_name beep.au
	*FvwmEvent: end_windowlist twang.au

	*FvwmEvent: icon_location beep.au
	*FvwmEvent: map beep.au
	*FvwmEvent: error beep.au
	*FvwmEvent: config_info beep.au
	*FvwmEvent: end_config_info beep.au
	*FvwmEvent: icon_file beep.au
	*FvwmEvent: default_icon beep.au
	*FvwmEvent: string plapper.au

	*FvwmEvent: mini_icon beep.au
	*FvwmEvent: windowshade beep.au
	*FvwmEvent: dewindowshade beep.au
.sp
.fi
Note: The enter_window event is generated when the pointer enters
a window.  With the -passid option, that window's id is passed to
fvwm.  An enter_window event is generated too when the pointer
leaves a window and moves into the root window.  In this case, the
id passed is 0.
.sp
.fi
Provided \fIfvwm\fP supports it (not yet), there's an additional event to
replace all \fIfvwm\fP beeps with a sound:
.nf
.sp
	*FvwmEvent: beep beep.au
.sp
.fi
The toggle_paging event will be supported, as soon, as it's
resurrected by \fIfvwm\fP:
.nf
.sp
	*FvwmEvent: toggle_paging fwop.au
.sp
.fi
.IP "*FvwmEvent: Delay \fI5\fP"
Specifies that an event-action will only be executed if it occurs at
least 5 seconds after the previous event.  Events that occur during
the delay period are ignored.  This option is useful if you don't want
several sounds playing at the same time.  The default delay is 0 which
disables the Event delay.

.IP "*FvwmEvent: StartDelay \fIdelay\fP"
Specifies that an event-action will only be executed if it occurs at
least \fIdelay\fP seconds after the startup event. Events that occur during
the delay period are ignored.  This option is useful when \fIfvwm\fP
starts and restarts using an audio player.  The default delay is 0.

.SH RPLAY OPTIONS
The following options are only valid with builtin rplay support.
i.e: when \fIFvwmEvent\fP was compiled with \fIHAVE_RPLAY\fP defined.
They are used only if \fIFvwmEvent: Cmd\fP is set
to \fIbuiltin-rplay\fP.


.IP "*FvwmEvent: RplayHost \fIhostname\fP"
Specifies what host the rplay sounds will play on.  The \fIhostname\fP
can also be an environment variable such as $HOSTDISPLAY.

.IP "*FvwmEvent: RplayPriority \fI0\fP"
Specifies what priority will be assigned to the rplay sounds when they
are played.

.IP "*FvwmEvent: RplayVolume \fI127\fP"
Specifies what volume will be assigned to the sounds when they are
played.

.SH FvwmAudio Compatibility Mode

When invoked in FvwmAudio compatibility mode (see above), FvwmEvent
accepts the following options to provide backwards compatibility
for FvwmAudio:

.IP "*FvwmEvent: PlayCmd \fIcommand\fP"
This is equivalent to using *FvwmEvent: Cmd to Exec commands. This
determines the independent audio player program that will actually
play the sounds. If the play command is set to \fIbuiltin-rplay\fP
then the builtin rplay support will be used.

.IP "*FvwmAudio: Dir \fIdirectory\fP"
Specifies the directory to look for the audio files.  This option is
ignored when rplay is used.

.SH BUGS
It's REALLY noisy when \fIfvwm\fP starts and restarts using an audio player.
You can use FvwmEvent: StartDelay to fix this problem.

.SH COPYRIGHTS
This module has evolved of \fIFvwmAudio\fP, which in term is heavily based
on a similar Fvwm module called \fIFvwmSound\fP by Mark
Boyns. \fIFvwmAudio\fP simply took Mark's original program and
extended it to make it generic enough to work with any Audio
player. Due to different requests to do specific things on specific events,
\fIFvwmEvent\fP took this one step further and now calls any
\fIfvwm\fP function, or builtin-rplay. If \fIfvwm\fP's Exec function
is used, any external program can be called with any parameter.

The concept for interfacing this module to the Window Manager, is
original work by Robert Nation.

Copyright 1998 Albrecht Kadlec.
Copyright 1994, Mark Boyns and Mark Scott.  No guarantees or
warranties or anything are provided or implied in any way whatsoever.
Use this program at your own risk.  Permission to use and modify this
program for any purpose is given, as long as the copyright is kept intact.


.sp
.SH AUTHORS
.nf
1994  FvwmSound  Mark Boyns       (\fIboyns@sdsu.edu\fP)
1994  FvwmAudio  Mark Scott       (\fImscott@mcd.mot.com\fP)
1996  FvwmAudio  Albrecht Kadlec
1998  FvwmEvent  Albrecht Kadlec  (\fIalbrecht@auto.tuwien.ac.at\fP)


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

modules/FvwmEvent/FvwmEvent.c
/* -*-c-*- */
/*
 * Copyright (C) 1994 Mark Boyns (boyns@sdsu.edu) and
 *                    Mark Scott (mscott@mcd.mot.com)
 *               1996-1998 Albrecht Kadlec (albrecht@auto.tuwien.ac.at)
 *
 * FvwmEvent version 1.0
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see: <http://www.gnu.org/licenses/>
 */

/* This module is based on FvwmModuleDebugger */

/* ---------------------------- included header files ----------------------- */

#include "config.h"

#include <stdio.h>
#include <time.h>

#include "libs/log.h"
#include "libs/Module.h"
#include "libs/fvwmlib.h"
#include "libs/Parse.h"
#include "libs/Strings.h"

/*
 * fvwm includes:
 */
#include <stdio.h>
#include <signal.h>
#include <sys/wait.h>
#include "libs/ftime.h"
#include <ctype.h>

/* ---------------------------- local definitions --------------------------- */

#define BUILTIN_STARTUP         0
#define BUILTIN_SHUTDOWN        1
#define BUILTIN_UNKNOWN         2
#define MAX_BUILTIN             3
#define SYNC_MASK_M             (M_DESTROY_WINDOW | M_LOWER_WINDOW | \
	M_RESTACK | M_CONFIGURE_WINDOW)
#define SYNC_MASK_MX            (M_EXTENDED_MSG)

/* ---------------------------- local macros -------------------------------- */

/* ---------------------------- imports ------------------------------------- */

/* ---------------------------- included code files ------------------------- */

/* ---------------------------- local types --------------------------------- */

typedef struct
{
	char *name;
	int action_arg;
	union
	{
		char *action;
	} action;
} event_entry;

/* ---------------------------- forward declarations ------------------------ */

void execute_event(event_entry*, short, unsigned long*);
void config(void);
RETSIGTYPE DeadPipe(int);
static RETSIGTYPE TerminateHandler(int);

/* ---------------------------- local variables ----------------------------- */

/* ---------------------------- exported variables (globals) ---------------- */

static char *MyName;
static int MyNameLen;
static int fd[2];
static char *cmd_line = NULL;
static time_t delay = 0; /* seconds */
static time_t last_time = 0;
static time_t now;
static time_t start_delay = 0;
/* don't tag on the windowID by default */
static Bool PassID = False;

#define ARG_NO_WINID 1024  /* just a large number */

#define EVENT_ENTRY(name,action_arg) { name, action_arg, {NULL} }
static event_entry message_event_table[] =
{
	EVENT_ENTRY( "new_page", -1 ),
	EVENT_ENTRY( "new_desk", 0 | ARG_NO_WINID ),
	EVENT_ENTRY( "old_add_window", 0 ),
	EVENT_ENTRY( "raise_window", 0 ),
	EVENT_ENTRY( "lower_window", 0 ),
	EVENT_ENTRY( "old_configure_window", 0 ),
	EVENT_ENTRY( "focus_change", 0 ),
	EVENT_ENTRY( "destroy_window", 0 ),
	EVENT_ENTRY( "iconify", 0 ),
	EVENT_ENTRY( "deiconify", 0 ),
	EVENT_ENTRY( "window_name", 0 ),
	EVENT_ENTRY( "icon_name", 0 ),
	EVENT_ENTRY( "res_class", 0 ),
	EVENT_ENTRY( "res_name", 0 ),
	EVENT_ENTRY( "end_windowlist", -1 ),
	EVENT_ENTRY( "icon_location", 0 ),
	EVENT_ENTRY( "map", 0 ),
	EVENT_ENTRY( "error", -1 ),
	EVENT_ENTRY( "config_info", -1 ),
	EVENT_ENTRY( "end_config_info", -1 ),
	EVENT_ENTRY( "icon_file", 0 ),
	EVENT_ENTRY( "default_icon", -1 ),
	EVENT_ENTRY( "string", -1 ),
	EVENT_ENTRY( "mini_icon", 0 ),
	EVENT_ENTRY( "windowshade", 0 ),
	EVENT_ENTRY( "dewindowshade", 0 ),
	EVENT_ENTRY( "visible_name", 0 ),
	EVENT_ENTRY( "sendconfig", -1 ),
	EVENT_ENTRY( "restack", -1 ),
	EVENT_ENTRY( "add_window", 0 ),
	EVENT_ENTRY( "configure_window", 0 ),
	EVENT_ENTRY(NULL,0)
};
static event_entry extended_message_event_table[] =
{
	EVENT_ENTRY( "visible_icon_name", 0 ),
	EVENT_ENTRY( "enter_window", 0 ),
	EVENT_ENTRY( "leave_window", 0 ),
	EVENT_ENTRY( "property_change", 0),
	EVENT_ENTRY( "monitor_enabled", 0 | ARG_NO_WINID),
	EVENT_ENTRY( "monitor_disabled", 0 | ARG_NO_WINID),
	EVENT_ENTRY( "monitor_changed", 0 | ARG_NO_WINID),
	EVENT_ENTRY( "monitor_focus", 0 | ARG_NO_WINID),
	EVENT_ENTRY( "echo", 0),
	EVENT_ENTRY( "reply", 0), /* FvwmEvent will never receive MX_REPLY */
	EVENT_ENTRY(NULL,0)
};
static event_entry builtin_event_table[] =
{
	EVENT_ENTRY( "startup", -1 ),
	EVENT_ENTRY( "shutdown", -1 ),
	EVENT_ENTRY( "unknown", -1),
	EVENT_ENTRY(NULL,0)
};

#if 0
/* These are some events described in the man page, which does not exist in
   fvwm. */
static event_entry future_event_table[] =
{
#ifdef M_BELL
	EVENT_ENTRY( "beep", -1 ),
#endif
	EVENT_ENTRY(NULL,0)
}
#endif

static event_entry* event_tables[] =
{
	message_event_table,
	extended_message_event_table,
	builtin_event_table,
	NULL
};

#undef EVENT_ENTRY

static unsigned int m_selected = 0;
static unsigned int mx_selected = 0;
static unsigned int m_sync = 0;
static unsigned int mx_sync = 0;


static volatile sig_atomic_t isTerminated = False;

/* ---------------------------- local functions ----------------------------- */

void unlock_event(unsigned long evtype)
{
	unsigned int mask;

	if (evtype & M_EXTENDED_MSG)
	{
		mask = mx_sync;
	}
	else
	{
		mask = m_sync;
	}
	if (evtype & mask)
	{
		SendUnlockNotification(fd);
	}

	return;
}

int main(int argc, char **argv)
{
	char *s;
	unsigned long header[FvwmPacketHeaderSize];
	unsigned long body[FvwmPacketBodyMaxSize];
	int total, remaining, count, event;
	int is_extended_msg;

	cmd_line = fxmalloc(1);
	*cmd_line = 0;
	/* Save our program  name - for error events */
	if ((s=strrchr(argv[0], '/')))
	{
		/* strip path */
		s++;
	}
	else
	{
		/* no slash */
		s = argv[0];
	}
	if (argc == 7)
	{
		/* use an alias */
		s = argv[6];
	}

	/* account for '*' */
	MyNameLen=strlen(s)+1;
	/* account for \0 */
	MyName = fxmalloc(MyNameLen+1);
	*MyName='*';
	/* append name */
	strcpy(MyName+1, s);

	/* Now MyName is defined */
	if ((argc != 6)&&(argc != 7))
	{
		fprintf(stderr, "%s Version "VERSION" should only be "
			   "executed by fvwm!\n", MyName+1);
		exit(1);
	}

#ifdef HAVE_SIGACTION
	{
		struct sigaction    sigact;

		sigemptyset(&sigact.sa_mask);
# ifdef SA_INTERRUPT
		sigact.sa_flags = SA_INTERRUPT;
# else
		sigact.sa_flags = 0;
# endif
		sigact.sa_handler = TerminateHandler;

		/* Dead pipe == Fvwm died */
		sigaction(SIGPIPE,&sigact,NULL);
		/* "polite" termination signal */
		sigaction(SIGTERM,&sigact,NULL);
	}
#else
	/* We don't have sigaction(), so fall back to less robust methods. */
	signal(SIGPIPE, TerminateHandler);
	signal(SIGTERM, TerminateHandler);
#endif

	fd[0] = atoi(argv[1]);
	fd[1] = atoi(argv[2]);

	/* configure events */
	config();
	/* Startup event */
	execute_event(builtin_event_table, BUILTIN_STARTUP, NULL);
	if (start_delay)
	{
		last_time = time(0);
	}
	/* tell fvwm we're running */
	SetMessageMask(fd, m_selected);
	SetMessageMask(fd, mx_selected | M_EXTENDED_MSG);
	m_sync = (m_selected & SYNC_MASK_M);
	mx_sync = (mx_selected & SYNC_MASK_M) | M_EXTENDED_MSG;
	/* migo (19-Aug-2000): synchronize on M_DESTROY_WINDOW
	 * dv (6-Jul-2002: synchronize on a number of events that can hide or
	 * destroy a window */
	if (m_sync)
	{
		SetSyncMask(fd, m_sync);
	}
	if (mx_sync != M_EXTENDED_MSG)
	{
		SetSyncMask(fd, mx_sync);
	}
	mx_sync &= ~M_EXTENDED_MSG;
	SendFinishedStartupNotification(fd);

	/* main loop */
	while (!isTerminated)
	{
		unsigned long msg_bit;
		event_entry *event_table;

		if ((count = read(fd[1], header, FvwmPacketHeaderSize_byte)) <=
		    0)
		{
			isTerminated = 1;
			continue;
		}
		/* if this read is interrrupted EINTR, the wrong event is
		 * triggered! */

		if (header[0] != START_FLAG)
		{
			/* should find something better for resyncing */
			unlock_event(header[1]);
			continue;
		}

		/* Ignore events that occur during the delay period. */
		now = time(0);

		/* junk the event body */
		total=0;
		remaining = (header[2] - FvwmPacketHeaderSize) *
			sizeof(unsigned long);
		while (remaining)
		{
			if ((count=read(fd[1],&body[total],remaining)) < 0)
			{
				isTerminated = 1;
				unlock_event(header[1]);
				continue;
			}
			remaining -= count;
			total +=count;
		}

		if (now < last_time + delay + start_delay)
		{
			/* quash event */
			unlock_event(header[1]);
			continue;
		}
		else
		{
			start_delay = 0;
		}

		/* event will equal the number of shifts in the base-2
		 * header[1] number.  Could use log here but this should be
		 * fast enough. */
		event = -1;
		msg_bit = header[1];
		is_extended_msg = (msg_bit & M_EXTENDED_MSG);
		msg_bit &= ~M_EXTENDED_MSG;
		while (msg_bit)
		{
			event++;
			msg_bit >>= 1;
		}
		if (is_extended_msg && event != -1)
		{
			event--;
		}
		if (
			event == -1 ||
			(event >= MAX_MESSAGES && !is_extended_msg) ||
			(event >= MAX_EXTENDED_MESSAGES && is_extended_msg))
		{
			event_table = builtin_event_table;
			event = BUILTIN_UNKNOWN;
		}
		else if (is_extended_msg)
		{
			event_table = extended_message_event_table;
		}
		else
		{
			event_table = message_event_table;
		}
		execute_event(event_table, event, body);
		unlock_event(header[1]);
	} /* while */
	execute_event(builtin_event_table, BUILTIN_SHUTDOWN, NULL);

	return 0;
}


/*
 *
 * Procedure:
 *
 *    execute_event - actually executes the actions from lookup table
 *
 */
void execute_event(event_entry *event_table, short event, unsigned long *body)
{
	if (event_table[event].action.action == NULL)
		return;

	char *buf = NULL;
	char *action;
	int action_arg = event_table[event].action_arg;
	int fw = 0;

	action = event_table[event].action.action;

	if (action_arg != -1 && !(action_arg & ARG_NO_WINID))
		fw = body[action_arg];

	if (PassID && action_arg != -1) {
		if (action_arg & ARG_NO_WINID) {
			action_arg &= ~ARG_NO_WINID;
			xasprintf(&buf, "%s %s %ld", cmd_line, action,
					body[action_arg]);
		} else {
			xasprintf(&buf, "%s %s 0x%lx", cmd_line, action,
					body[action_arg]);
		}
	} else {
		char *mon_event = "";
		char *extra_arg = "";

		if (StrHasPrefix(event_table[event].name, "monitor_"))
			mon_event = event_table[event].name;
		extra_arg = (char *)&body[3];

		xasprintf(&buf, "%s %s %s %s", cmd_line, action, mon_event,
				extra_arg);
	}

	/* let fvwm execute the function */
	SendText(fd, buf, fw);
	last_time = now;

	free(buf);

	return;
}


/*
 *
 *  Procedure:
 *      config - read the configuration file.
 *
 */

void handle_config_line(char *buf)
{
	char *event;
	char *action;
	char *p;
	char *token;
	char **e;
	int i;
	int found;

	char *table[]=
	{
		"Cmd",
		"Delay",
		"PassID",
		"StartDelay"
	}; /* define entries here, if this list becomes unsorted, use FindToken */


	p = buf + MyNameLen;
	/* config option ? */
	if ((e = FindToken(p, table, char *)))
	{
		/* skip matched token */
		p += strlen(*e);
		token = PeekToken(p, &p);
		switch (e - (char**)table)
		{
		case 0:
			/* Cmd */
			if (cmd_line)
			{
				free(cmd_line);
			}
			if (token)
			{
				cmd_line = fxstrdup(token);
			}
			else
			{
				cmd_line = fxstrdup("");
			}
			break;
		case 1:
			/* Delay */
			if (token)
			{
				delay = atoi(token);
			}
			break;
		case 2:
			/* PassID */
			PassID = True;
			break;
		case 3:
			/* StartDelay */
			if (token)
			{
				start_delay = atoi(token);
			}
			break;
		}
	}
	else
	{
		event_entry **event_table;

		/* test for isspace(*p) ??? */
		p = GetNextSimpleOption(p, &event);
		p = GetNextSimpleOption(p, &action);
		if (!event || !*event || !action || !*action)
		{
			if (event)
			{
				free(event);
			}
			if (action)
			{
				free(action);
			}
			fprintf(stderr, "%s: incomplete event definition %s\n",
				   MyName + 1, buf);
			return;
		}
		event_table = event_tables;
		found = 0;
		while(found == 0 && *event_table != NULL)
		{
			event_entry *loop_event;

			for (loop_event = *event_table, i = 0;
			     found == 0 && loop_event->name != NULL;
			     loop_event++, i++)
			{
				if (MatchToken(event, loop_event->name))
				{
					if (loop_event->action.action != NULL)
					{
						free(loop_event->action.action);
					}
					loop_event->action.action = action;
					found = 1;
					if (*event_table == message_event_table)
					{
						m_selected |= (1 << i);
					}
					else if (
						*event_table ==
						extended_message_event_table)
					{
						mx_selected |= (1 << (i + 1));
					}
				}
			}
			event_table++;
		}
		if (!found)
		{
			fprintf(stderr, "%s: unknown event type: %s\n",
				   MyName+1, event);
			if (action)
			{
				free(action);
			}
		}
		if (event)
		{
			free(event);
		}
	}

	return;
}

void config(void)
{
	char *buf;

	/* get config lines with my name */
	InitGetConfigLine(fd,MyName);
	while (GetConfigLine(fd,&buf), buf != NULL && *buf != 0)
	{
		if (buf[strlen(buf)-1] == '\n')
		{
			/* if line ends with newline, strip it off */
			buf[strlen(buf)-1] = '\0';
		}

		/* Search for MyName (normally *FvwmEvent) */
		if (strncasecmp(buf, MyName, MyNameLen) == 0)
		{
			handle_config_line(buf);
		}
	}
}

/*
 *
 *  Procedure:
 *      SIGPIPE handler - SIGPIPE means fvwm is dying
 *
 */
static RETSIGTYPE
TerminateHandler(int nonsense)
{
	isTerminated = True;

	SIGNAL_RETURN;
}

/*
 *
 *  Procedure:
 *      Externally callable procedure to quit
 *
 */
RETSIGTYPE DeadPipe(int flag)
{
	execute_event(builtin_event_table, BUILTIN_SHUTDOWN, NULL);
	exit(flag);
	SIGNAL_RETURN;
}
